[SPARK-53207][SDP] Send Pipeline Event to Client Asynchronously

JiaqiWang18 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Created `PipelineEventSender` to allow sending pipeline event back to client in a background thread to not block pipeline execution. New events gets queued into the sender and will get processed sequentially. The sender waits until all events are sent back to client before shutting down the processing loop.

Implemented logical capacity checking that examines event types before submission, ensuring RunProgress and terminal FlowProgress events are always queued while other events may be dropped when the queue is full. This prevents buffer from overflowing and impact pipeline execution when the number of events are too large.

The queue size can be controlled by a spark conf `spark.sql.connect.pipeline.event.queue.capacity`, currently defaulted to `1000`


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
To ensure event sending do not block pipeline execution.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as new features, bug fixes, or other behavior changes. Documentation-only updates are not considered user-facing changes.

If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
New and existing tests to ensure events are delivered and not done in a blocking way.


### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

jackywang-db committed:
initial impl

jackywang-db committed:
fmt

jackywang-db committed:
await on shutdown

jackywang-db committed:
test

jonmio commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,194 @@
+import org.apache.spark.sql.pipelines.logging.PipelineEvent
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * Handles sending pipeline events to the client in a background thread using ExecutorService.
+ * This prevents pipeline execution from blocking on response observer operations and provides
+ * better error isolation between pipeline execution and client communication.
+ *
+ * Based on the executor pattern used in DataPlaneEventLog for better thread management.
+ */
```
```suggestion
/**
 * Handles sending pipeline events to the client in a background thread.
 * This prevents pipeline execution from blocking on streaming events.
 */
```

jonmio replied:
Super nit: The references to specfic class names are not super helpful in this case, and some of those references don't exist in this repo

jonmio commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,194 @@
+  }
+
+  // state tracking
+  private val isShutdown = new AtomicBoolean(false)
+  private val isStarted = new AtomicBoolean(false)
```
Let's document how/when these state transitions happen and what these flags are used for

jonmio commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,194 @@
+  def sendEvent(event: PipelineEvent): Unit = {
+    if (!isShutdown.get()) {
+      val offered = eventQueue.offer(EventMessage(event))
+      if (!offered) {
+        logWarning(s"Failed to queue pipeline event for session ${sessionHolder.sessionId}")
```
When might this happen?

jonmio commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,194 @@
+   * Send an event asynchronously by adding it to the queue
+   */
+  def sendEvent(event: PipelineEvent): Unit = {
+    if (!isShutdown.get()) {
```
What is the reason behind this check?

jonmio commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,194 @@
+        message match {
+          case EventMessage(event) =>
+            sendEventToClient(event)
+          case ShutdownMessage =>
```
Do we really need this ShutdownMessage? Can the shutdown call not just interrupt this thread?

JiaqiWang18 committed:
Update sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala

Co-authored-by: Jon Mio <jon7470@gmail.com>

jackywang-db committed:
address feedback

jackywang-db committed:
fmt

jackywang-db committed:
use ThreadUtils.newDaemonSingleThreadExecutor

jackywang-db committed:
use AutoCloseable to avoid manual shutdown() call

JiaqiWang18 commented:
* refactored to use `ThreadUtils.newDaemonSingleThreadExecutor` to avoid the need of maintaining a queue manually. Each event is queued as a runnable task that sends the event.
* implemented the `AutoCloseable` interface with try-with-resources to avoid the need of manual `shutdown()` call to avoid the caller forgetting, which could cause events to be lost.

@jonmio could you take another pass?

jackywang-db committed:
build

jackywang-db committed:
bounded queue

jackywang-db committed:
Logical queue size with shouldEnqueueEvent

JiaqiWang18 commented:
@anew  @anishm-db Added queue capacity limit with logical capacity checking to ensure all `RunProgress` events are queued and sent and only terminal `FlowProgress` events are queued and sent when reaching the limit. Logic is in `shouldEnqueueEvent`

jackywang-db committed:
nit

jackywang-db committed:
nit

jackywang-db committed:
use SQLConf

anew reviewed (approved):
this looks good to me. Did not review tests in details.

jackywang-db committed:
use structured logging

gengliangwang commented on sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala:
```diff
@@ -0,0 +1,185 @@
+                      // lost or double-counted.
+                      .setSeconds(event.timestamp.getTime / 1000)
+                      .setNanos(event.timestamp.getNanos)
+                      .build())
```
nit: let's avoid building so many layers of builder in on function call.

JiaqiWang18 replied:
@gengliangwang extracted a helper [constructProtoEvent](https://github.com/apache/spark/pull/51956/commits/2bf017c1290470ccf3d94816dc851f07c00b3620) to reduce the nesting

gengliangwang reviewed (approved):
LGTM except for one comment

jackywang-db committed:
extract constructProtoEvent

gengliangwang commented:
Thanks, merging to master
